src/ModelContextProtocol.AspNetCore/McpEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,108 @@
+/// </summary>
+public static class McpEndpointRouteBuilderExtensions
+{
+    /// <summary>
+    /// Sets up endpoints for handling MCP "HTTP with SSE" transport.
+    /// </summary>
+    /// <param name="endpoints">The web application to attach MCP HTTP endpoints.</param>
+    /// <param name="onSessionAsync">Provides an optional asynchronous callback for handling new MCP sessions.</param>
+    /// <returns>Returns a builder for configuring additional endpoint conventions like authorization policies.</returns>
+    public static IEndpointConventionBuilder MapMcp(this IEndpointRouteBuilder endpoints, Func<HttpContext, IMcpServer, Task>? onSessionAsync = null)
+    {
+        ConcurrentDictionary<string, SseResponseStreamTransport> _sessions = new(StringComparer.Ordinal);
+
+        var loggerFactory = endpoints.ServiceProvider.GetRequiredService<ILoggerFactory>();
+        var mcpServerOptions = endpoints.ServiceProvider.GetRequiredService<IOptions<McpServerOptions>>();
+
+        var routeGroup = endpoints.MapGroup("");
+
+        routeGroup.MapGet("/sse", async context =>
+        {
+            var response = context.Response;
+            var requestAborted = context.RequestAborted;
+
+            response.Headers.ContentType = "text/event-stream";
+            response.Headers.CacheControl = "no-cache";
+
+            var sessionId = MakeNewSessionId();
+            await using var transport = new SseResponseStreamTransport(response.Body, $"/message?sessionId={sessionId}");
+            await using var server = McpServerFactory.Create(transport, mcpServerOptions.Value, loggerFactory, endpoints.ServiceProvider);
+            _sessions.TryAdd(sessionId, transport);
+
+            try
+            {
+                var transportTask = transport.RunAsync(cancellationToken: requestAborted);
+                var serverTask = server.RunAsync(cancellationToken: requestAborted);
+                if (onSessionAsync is not null)
+                {
+                    await onSessionAsync(context, server);
+                }
+                await serverTask;
+                await transportTask;
+            }
+            catch (OperationCanceledException) when (requestAborted.IsCancellationRequested)
+            {
+                // RequestAborted always triggers when the client disconnects before a complete response body is written,
+                // but this is how SSE connections are typically closed.
+            }
+            finally
+            {
+                _sessions.TryRemove(sessionId, out _);
+            }
+        });
+
+        routeGroup.MapPost("/message", async context =>
+        {
+            if (!context.Request.Query.TryGetValue("sessionId", out var sessionId))
+            {
+                await Results.BadRequest("Missing sessionId query parameter.").ExecuteAsync(context);
+                return;
+            }
+
+            if (!_sessions.TryGetValue(sessionId.ToString(), out var transport))
```

Comment:
What is sessionId here... a StringValues that in the 99.999% case has just a single value?

Reply:
Correct. If there's something like `?sessionId=123&sessionId=345", StringValues.ToString() will concantinate that into "123,345" which won't match any session IDs since they never contain commas. We could check that there's a single value first, but that seems like a super premature optimization.